### ISSUE TITLE
MaxPool2d returns FloatTensor as indices

### ISSUE THREAD

**Issue by magnusja** [pytorch/pytorch#7336]
## Issue description

MaxPool2d returns a FloatTensor as indices which results in an exception with MaxUnpool2d.

## Code example

```
    def forward(self, input):
        tmp = self.basic(input)
        out, indices = self.pool(tmp) # indices are FloatTensor (sometimes)
        return out, indices, tmp

      def forward(self, input, indices, encoder_block):
        tmp = self.unpool(input, indices) # exception expected LongTensor for indices
        tmp = torch.cat((encoder_block, tmp), dim=1)
        return self.basic(tmp)
```

## System Info

```
Collecting environment information...
PyTorch version: 0.4.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 17.10
GCC version: (Ubuntu 7.2.0-8ubuntu3.2) 7.2.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5

Versions of relevant libraries:
[pip3] numpy (1.14.2)
[pip3] torch (0.4.0)
[pip3] torchvision (0.2.1)
[conda] Could not collect
```

**Comment by li-roy** [comment]
Looks like you didn't set `return_indices=True` on MaxPool2D. In most cases, this would throw an error when unpacking `out, indices = self.pool(tmp)`, but if the output's 0th dim has size two it wouldn't throw.

**Comment by magnusja** [comment]
This is self.pool:

```
self.pool = nn.MaxPool2d(kernel, stride_pool, return_indices=True)
```

The strange thing is that it returns this FloatTensor only sometimes. The only similar issue I could find was this one: https://github.com/pytorch/pytorch/issues/1197

I have this problem when using cuda as well as CPU. Do you maybe have any hints how to debug this?

**Comment by magnusja** [comment]
Hey guys,

so I confused index and input somewhere in my code. But what was confusing that if I print indices they sometimes are of type float. But maybe that is alright like in https://github.com/pytorch/pytorch/issues/1197

So sorry that I bothered you with that!
